workflow/algorithm.go

```diff
@@ -0,0 +1,98 @@
+package workflow
+
+import "fmt"
+
+// ChildrenIDs returns the list of node IDs with a dependency to the current node
+func (w Workflow) ChildrenIDs(nodeKey string) []string {
+	nodeKeys := make([]string, 0)
+	for _, edge := range w.Edges {
+		if edge.Src == nodeKey {
+			nodeKeys = append(nodeKeys, edge.Dst)
+		}
+	}
+	return nodeKeys
+}
+
+// ParentIDs returns the list of node IDs with the current node as child
+func (w Workflow) ParentIDs(nodeKey string) []string {
+	nodeKeys := make([]string, 0)
+	for _, edge := range w.Edges {
+		if edge.Dst == nodeKey {
+			nodeKeys = append(nodeKeys, edge.Src)
+		}
+	}
+	return nodeKeys
+}
+
+// FindNode returns the node matching the key in parameter or an error if not found
+func (w Workflow) FindNode(key string) (Node, error) {
+	for _, node := range w.Nodes {
+		if node.Key == key {
+			return node, nil
+		}
+	}
+	return Node{}, fmt.Errorf("node %q not found", key)
+}
+
+// A null graph is a graph that contains no nodes
+func (w *Workflow) isNullGraph() bool {
+	if w.Nodes == nil {
+		return true
+	}
+	return len(w.Nodes) == 0
+}
+
+// An acyclic graph is a graph that doesn't contain a cycle. If you walk through the graph you will go maximum one time on each node.
+func (w *Workflow) isAcyclic(node string, discovered map[string]bool) bool {
+	if discovered[node] {
+		return false
+	}
+	discovered[node] = true
+	for _, child := range w.ChildrenIDs(node) {
+		copyMap := make(map[string]bool)
+		for k, v := range discovered {
+			copyMap[k] = v
+		}
+		if !w.isAcyclic(child, copyMap) {
+			return false
+		}
+	}
+	return true
+}
+
+// A connected graph is a graph where all the nodes are connected with each other through edges.
+// Warning: this function will have a stack overflow if the graph is not acyclic.
+func (w *Workflow) isConnected(node string) bool {
```

Comment:
```suggestion
func (w *Workflow) isConnected() bool {
```

It should retruns if a graph is connected without specifying the starting node. the docs says that the method returns true if all the nodes are connected.